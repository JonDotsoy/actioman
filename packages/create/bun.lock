{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "create-actioman",
      "dependencies": {
        "@jondotsoy/flags": "^2.3.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@jondotsoy/console-draw": ["@jondotsoy/console-draw@0.1.2", "", { "dependencies": { "@jondotsoy/style-text": "^0.9.1" }, "peerDependencies": { "prettier": "^3.3.3", "typescript": "^5.0.0" } }, "sha512-BBZGLd6RzjKxKcgD2OwD7v6GeKlf53Ysk59H+7+HvyGLcJZ7AJk8rOCq/yv813KbWV2teTCEBQXnnP5hZe/2CA=="],

    "@jondotsoy/flags": ["@jondotsoy/flags@2.3.0", "", { "dependencies": { "@jondotsoy/console-draw": "^0.1.2" } }, "sha512-vUMe4NdXKv8RxMXrKCJLsHBt3XQ8/YV4Fyr1I43vSLe5v0M8sdY3wJl1qdTgwRBZES2zMLoenxOUSNGE9r/5NA=="],

    "@jondotsoy/style-text": ["@jondotsoy/style-text@0.9.2", "", { "peerDependencies": { "typescript": "^5.0.0" } }, "sha512-6fonOzrANwrw+UIM4fivtDZLvuCN00yH0lCEYi14ln4j8fSHZokb6TLrqOu38GCqqQnZ2LuzJTQBWgv7jN6iVA=="],

    "@types/bun": ["@types/bun@1.2.2", "", { "dependencies": { "bun-types": "1.2.2" } }, "sha512-tr74gdku+AEDN5ergNiBnplr7hpDp3V1h7fqI2GcR/rsUaM39jpSeKH0TFibRvU0KwniRx5POgaYnaXbk0hU+w=="],

    "@types/node": ["@types/node@22.13.1", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-jK8uzQlrvXqEU91UxiK5J7pKHyzgnI1Qnl0QDHIgVGuolJhRb9EEl28Cj9b3rGR8B2lhFCtvIm5os8lFnO/1Ew=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.2", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-RCbMH5elr9gjgDGDhkTTugA21XtJAy/9jkKe/G3WR2q17VPGhcquf9Sir6uay9iW+7P/BV0CAHA1XlHXMAVKHg=="],

    "prettier": ["prettier@3.4.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-e9MewbtFo+Fevyuxn/4rrcDAaq0IYxPGLvObpQjiZBMAzB9IGmzlnG9RZy3FFas+eBMu2vA0CszMeduow5dIuQ=="],

    "typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
