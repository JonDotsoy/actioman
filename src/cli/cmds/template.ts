import {
  flag,
  flags,
  isBooleanAt,
  makeHelpMessage,
  rule,
  type Rule,
} from "@jondotsoy/flags";

export const template = async (args: string[]) => {
  type Options = { help: boolean };
  const rules: Rule<Options>[] = [
    rule(flag("-h", "--help"), isBooleanAt("help"), {
      description: "Show help",
    }),
  ];
  const options = flags(args, {}, rules);

  if (options.help) return makeHelpMessage("actioman", rules);
};
