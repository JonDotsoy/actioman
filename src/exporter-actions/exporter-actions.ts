import { Actions } from "../actions/actions.js";
import zodToJsonSchema from "zod-to-json-schema";
import type { ActionsDefinitionsJsonDTO } from "./dtos/actions-definitions-json.dto.js";
import { get } from "@jondotsoy/utils-js/get";
import jsonSchemaToZod from "json-schema-to-zod";
import { actionDocumentTemplate } from "../gens-templates/action-document.template.js";

/** @deprecated by {@link actionDocumentTemplate} */
const templateAction = (
  actionNameSrc: string,
  descriptionSrc: string,
  inputSrc: string,
  outputSrc: string,
) => `    ${actionNameSrc}: {
      description: ${descriptionSrc},
      input: ${inputSrc},
      output: ${outputSrc},
    },`;

/** @deprecated by {@link actionDocumentTemplate} */
const template = (
  targetUrlSrc: string,
  actionsSrc: string,
  actionsTargetModuleLocation: string,
) => `import { z } from "zod";
import { ActionsTarget } from "${actionsTargetModuleLocation}";

const createActionsTarget = () =>
  new ActionsTarget(${targetUrlSrc}, {
${actionsSrc}
  }).compile();

export default createActionsTarget;
`;

type ToStringOptions = {
  fileLocation?: string;
  actionTargetModuleLocation?: string;
};

export class ActionsDocument {
  constructor(
    readonly targetUrl: URL,
    readonly actionsJson: any,
    readonly fileLocation?: string,
    readonly actionsTargetModuleLocation: string = "../../actions-target/actions-target.js",
  ) {}

  toString(options?: ToStringOptions) {
    const fileLocation =
      options?.fileLocation ?? this.fileLocation ?? process.cwd();
    const actionTargetModuleLocation =
      options?.actionTargetModuleLocation ?? this.actionsTargetModuleLocation;
    const toJsonZod = (value: Record<any, any> | undefined) => {
      if (value) return jsonSchemaToZod(value);
      return "undefined";
    };

    const actionsSrc = Array.from(
      Object.entries(this.actionsJson),
      ([name, actionDescription]) =>
        templateAction(
          JSON.stringify(name),
          JSON.stringify(get.string(actionDescription, "description")),
          toJsonZod(get.record(actionDescription, "input")),
          toJsonZod(get.record(actionDescription, "output")),
        ),
    ).join("\n");

    const targetUrlSrc = JSON.stringify(this.targetUrl.toString());

    return actionDocumentTemplate({
      fileLocation: fileLocation,
      actionTargetModuleLocation,
      actionsDocument: this,
    });
    // return template(targetUrlSrc, actionsSrc, this.actionsTargetModuleLocation);
  }

  static async fromHTTPServer(
    url: URL,
    cwd?: string,
    actionsTargetModuleLocation?: string,
  ) {
    const targetUrl = new URL("./__actions", url);
    const res = await fetch(targetUrl);
    const b = await res.json();
    const actionsJson = b.actions;
    return new ActionsDocument(
      targetUrl,
      actionsJson,
      cwd,
      actionsTargetModuleLocation,
    );
  }
}

export const actionsToJson = (actions: Actions): ActionsDefinitionsJsonDTO => {
  return Object.fromEntries(
    Array.from(
      Object.entries(Actions.describe(actions)),
      ([name, definition]) => [
        name,
        {
          description: definition.description ?? null,
          sse: definition.sse,
          input: definition.input
            ? zodToJsonSchema(definition.input, { target: undefined })
            : null,
          output: definition.output
            ? zodToJsonSchema(definition.output, { target: undefined })
            : null,
        },
      ],
    ),
  );
};

/** @deprecated by {@link ActionsDocument} */
export const importActionsFromJson = (
  json: any,
  target: string,
  uids = new Map<string, number>(),
) => {
  const id = (name: string) => {
    const index = uids.get(name) ?? 0;
    uids.set(name, index + 1);
    if (index === 0) return name;
    return `${name}_${index}`;
  };

  function* compile() {
    const definitions = get.record(json);
    if (!definitions) return null;
    for (const [name, definition] of Object.entries(definitions)) {
      const nameId = id(name);
      const description = get.string(definition, "description");
      const input: any = get(definition, "input");
      const output: any = get(definition, "output");

      const inputZodSrc = input ? jsonSchemaToZod(input, {}) : null;
      const outputZodSrc = output ? jsonSchemaToZod(output, {}) : null;

      const sources = {
        parts: {
          input: { js: inputZodSrc },
          output: { js: outputZodSrc },
        },
        chunkModule: {
          ts:
            "" +
            `export const ${nameId} = {\n` +
            `  description: ${JSON.stringify(description)},\n` +
            `  target: ${JSON.stringify(new URL(`./__actions/${name}`, target).toString())}\n` +
            `  input: ${inputZodSrc},\n` +
            `  output: ${outputZodSrc},\n` +
            `} as const;\n` +
            "\n",
        },
      };

      yield {
        documentId: nameId,
        name,
        description,
        sources,
      };
    }
  }

  const compiled = Array.from(compile());

  return (
    "" +
    `import { z } from "zod";\n` +
    "\n" +
    compiled
      .map(
        ({
          sources: {
            chunkModule: { ts },
          },
        }) => ts,
      )
      .join("") +
    ""
  );
};
