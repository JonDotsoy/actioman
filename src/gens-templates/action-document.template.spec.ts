import { describe, expect, it } from "bun:test";
import { actionDocumentTemplate } from "./action-document.template";
import { ActionsDocument } from "../exporter-actions/exporter-actions";
import zodToJsonSchema from "zod-to-json-schema";
import { z } from "zod";

describe("actionDocumentTemplate", () => {
  it("should generate the actions target code", () => {
    expect(
      actionDocumentTemplate({
        cwd: "file:///app/my_app/my_actions/",
        actionTargetModuleLocation:
          "file:///app/my_app/node_modules/actioman/lib/esm/actions-target/actions-target.js",
        actionsDocument: new ActionsDocument(new URL("http://localhost"), {
          hi: {
            description: "foo",
            input: zodToJsonSchema(
              z.object({
                name: z.string(),
              }),
            ),
            output: zodToJsonSchema(
              z.object({
                outname: z.number().nullable(),
              }),
            ),
          },
          foo: {
            description: "foo",
            input: undefined,
            output: zodToJsonSchema(
              z.object({
                outname: z.number().nullable(),
              }),
            ),
          },
        }),
      }),
    ).toMatchSnapshot();
  });
});
