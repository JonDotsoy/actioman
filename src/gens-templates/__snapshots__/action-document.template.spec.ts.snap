// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`actionDocumentTemplate should generate the actions target code 1`] = `
"// @ts-nocheck
import { z } from "zod";
import { ActionsTarget } from "./../node_modules/actioman/lib/esm/actions-target/actions-target.js";
const createActionsTarget = () => new ActionsTarget("http://localhost/", {
    hi: {
        description: "foo",
        input: z.object({ "name": z.string() }).strict(),
        output: z.object({ "outname": z.union([z.number(), z.null()]) }).strict()
    },
    foo: {
        description: "foo",
        input: undefined,
        output: z.object({ "outname": z.union([z.number(), z.null()]) }).strict()
    }
}).compile();
export default createActionsTarget;
"
`;
